text/0000-eager-macro-expansion.md

```diff
@@ -0,0 +1,615 @@
+    // to do so.
+    ```
+
+* In procedural macros:
+    ```rust
+    my_proc_macro!(concat!("hello", "world"));
+    //             ^^^^^^^^^^^^^^^^^^^^^^^^^
+    // This call isn't expanded before being passed to `my_proc_macro`, and
+    // can't be since proc macros are passed opaque token streams by design.
+    ```
+
+* In attribute macros:
+    ```rust
+    #[my_attr_macro(x = a_macro_call!(...))]
+    //                  ^^^^^^^^^^^^^^^^^^
+    // This call isn't expanded before being passed to `my_attr_macro`, and
+    // can't be since attr macros are passed opaque token streams by design.
+    struct X {
+        my_field_definition_macro!(...)
+    //  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Same with this one.
+    }
+    ```
+
+In these situations, macros need to either re-emit the input macro invocation
+as part of their token output, or simply reject the input. If the proc macro
+needs to inspect the result of the macro call (for instance, to check or edit
+it, or to re-export a hygienic symbol defined in it), the author is currently
+unable to do so.
+
+Giving proc macro authors the ability to handle these situations will allow
+proc macros to 'just work' in more contexts, and without surprising users who
+expect macro calls to interact well with more parts of the language.
+
+As a side note, allowing macro calls in built-in attributes would solve a few
+outstanding issues (see
+[rust-lang/rust#18849](https://github.com/rust-lang/rust/issues/18849) for an
+example).
+
+An older motivation to allow macro calls in attributes was to get
+`#[doc(include_str!("path/to/doc.txt"))]` working, in order to provide an
+ergonomic way to keep documentation outside of Rust source files. This was
+eventually emulated by the accepted [RFC
+1990](https://github.com/rust-lang/rfcs/pull/1990), indicating that macros in
+attributes could be used to solve problems at least important enough to go
+through the RFC process.
+
+## Interpolating macros in output
+
+Macros are currently not allowed in certain syntactic positions. Famously, they
+aren't allowed in identifier position, which makes `concat_idents!` [almost
+useless](https://github.com/rust-lang/rust/issues/29599). If macro authors have
+access to eager expansion, they could eagerly expand `concat_idents!` and
+interpolate the resulting token into their output.
+
+# Guide-level explanation
+
+We expose the following API in the `proc_macro` crate, to allow proc macro
+authors to iteratively expand all macros within a token stream:
+
+```rust
+use proc_macro::{LexError, Diagnostic, TokenStream};
+
+pub struct ExpansionBuilder { /* No public fields. */ }
+pub enum ExpansionError {
+    LexError(LexError),
+    ParseError(Diagnostic),
+    MissingDefinition(Diagnostic),
+    /* What other errors do we want to report?
+       What failures can the user reasonably handle? */
+}
+
+impl ExpansionBuilder {
+    /// Creates a new macro expansion request to iteratively expand all macro
```

Comment:
I think it would be more useful to be able to expand invocations one 'layer' at a time (e.g. expand all invocations in the `TokenStream`, but don't recursively expand any new invocations that appear). This would allow proc macros to avoid unnecessary work when they only care about the result of a particular expansion.

For example, a macro might want to evaluate something like `quote! { some_macro!($user_input) }` purely for its side effects (if `some_macro!` might emit an error), and not actually care about expanding any other macros that occur in the output of `some_macro!`.

An 'iterative expansion' method could be built on top of this, possibly as a helper method that avoids repeatedly `.await`ing the result.

Reply:
I think this would have limited usefulness _in practice_ - plenty of decl macros have a large number of intermediate expansions - but I agree that there's nothing objectionable in exposing a 'single step' expansion function. I'll update the RFC.